File: ea-adapter/src/de/proskor/automation/impl/NodeImpl.scala
package de.proskor.automation.impl

import de.proskor.automation.Node
import de.proskor.automation.Diagram
import de.proskor.automation.Element
import de.proskor.automation.Repository
import cli.EA.IDiagramObject

class NodeImpl(private[automation] val peer: IDiagramObject) extends Node {
  override def left: Int = peer.get_left
  override def left_=(left: Int) {
    peer.set_right(peer.get_right - peer.get_left + left)
    peer.set_left(left)
    peer.Update()
  }

  override def top: Int = -peer.get_top
  override def top_=(top: Int) {
    peer.set_bottom(peer.get_bottom - peer.get_top - top)
    peer.set_top(-top)
    peer.Update()
  }

  override def width: Int = peer.get_right - peer.get_left
  override def width_=(width: Int) {
    peer.set_right(peer.get_left + width)
    peer.Update()
  }

  override def height: Int = peer.get_top - peer.get_bottom
  override def height_=(height: Int) {
    peer.set_bottom(peer.get_top - height)
    peer.Update()
  }

  override def sequence: Int = peer.get_Sequence.asInstanceOf[Int]
  override def sequence_=(sequence: Int) {
    peer.set_Sequence(sequence)
    peer.Update()
  }

  override def diagram: Diagram = new DiagramImpl(RepositoryImpl.getDiagramById(peer.get_DiagramID))

  override def element: Element = new ElementImpl(RepositoryImpl.getElementById(peer.get_ElementID))
  override def element_=(element: Element) = peer.set_ElementID(element.id)

  override def equals(that: Any): Boolean = that match {
    case node: Node => element == node.element && diagram == node.diagram
    case _ => false
  }

  override def hashCode: Int = element.id
}

File: ea-automation/src/de/proskor/model/impl/NodeImpl.java
package de.proskor.model.impl;

import cli.EA.ICollection;
import cli.EA.IDiagram;
import cli.EA.IDiagramObject;
import cli.EA.IRepository;
import de.proskor.model.Diagram;
import de.proskor.model.Element;
import de.proskor.model.Node;

class NodeImpl extends IdentityImpl<IDiagramObject> implements Node {
	private final int diagramId;

	public NodeImpl(IRepository repository, int diagramId, int nodeId) {
		super(repository, nodeId);
		this.diagramId = diagramId;
	}

	@Override
	protected IDiagramObject getPeer() {
		final IDiagram diagram = (IDiagram) this.getRepository().GetDiagramByID(this.diagramId);
		final ICollection collection = (ICollection) diagram.get_DiagramObjects();
		final short count = collection.get_Count();
		for (short i = 0; i < count; i++) {
			final IDiagramObject node = (IDiagramObject) collection.GetAt(i);
			if (node.get_InstanceID() == this.getId())
				return node;
		}
		throw new IllegalStateException();
	}

	@Override
	public Element getElement() {
		final IDiagramObject peer = this.getPeer();
		final int elementId = peer.get_ElementID();
		return new ElementImpl(this.getRepository(), elementId);
	}

	@Override
	public Diagram getDiagram() {
		return new DiagramImpl(this.getRepository(), this.diagramId);
	}

	@Override
	public int getLeft() {
		final IDiagramObject peer = this.getPeer();
		return peer.get_left();
	}

	@Override
	public void setLeft(int left) {
		final IDiagramObject peer = this.getPeer();
		peer.set_left(peer.get_right() - peer.get_left() + left);
	}

	@Override
	public int getTop() {
		final IDiagramObject peer = this.getPeer();
		return -peer.get_top();
	}

	@Override
	public void setTop(int top) {
		final IDiagramObject peer = this.getPeer();
		peer.set_top(peer.get_bottom() - peer.get_top() - top);
	}

	@Override
	public int getRight() {
		final IDiagramObject peer = this.getPeer();
		return peer.get_right();
	}

	@Override
	public void setRight(int right) {
		final IDiagramObject peer = this.getPeer();
		peer.set_right(right);
	}

	@Override
	public int getBottom() {
		final IDiagramObject peer = this.getPeer();
		return -peer.get_bottom();
	}

	@Override
	public void setBottom(int bottom) {
		final IDiagramObject peer = this.getPeer();
		peer.set_bottom(-bottom);
	}

	@Override
	public int getSequence() {
		final IDiagramObject peer = this.getPeer();
		return Integer.valueOf(peer.get_Sequence().toString());
	}

	@Override
	public void setSequence(int sequence) {
		final IDiagramObject peer = this.getPeer();
		peer.set_Sequence(sequence);
	}

	@Override
	public String toString() {
		return "NODE[" + this.getId() + "] FOR " + this.getElement().toString();
	}

	@Override
	public boolean equals(Object that) {
		if (that == null)
			return false;

		if (that instanceof NodeImpl)
			return this.getId() == ((NodeImpl) that).getId();

		return false;
	}

	@Override
	public int hashCode() {
		return this.getId();
	}
}
